pkg/util/ip_reservation_test.go

```diff
@@ -0,0 +1,373 @@
+			expected:      true,
+			errorExpected: false,
+		},
+		{
+			name:          "Non overlapping CIDRs",
+			cidr1:         "192.168.92.0/29",
+			cidr2:         "192.168.22.67/26",
+			expected:      false,
+			errorExpected: false,
+		},
+		{
+			name:          "Non overlapping CIDRs with same cidr size",
+			cidr1:         "192.168.92.247/29",
+			cidr2:         "192.168.92.248/29",
+			expected:      false,
+			errorExpected: false,
+		},
+		{
+			name:          "Overlapping CIDRs with same cidr size",
+			cidr1:         "192.168.92.249/29",
+			cidr2:         "192.168.92.255/29",
+			expected:      true,
+			errorExpected: false,
+		},
+		{
+			name:          "Invalid CIDR provided",
+			cidr1:         "192.168.92.0",
+			cidr2:         "192.168.22.67/26",
+			errorExpected: true,
+		},
+	}
+
+	for _, test := range cases {
+		_, ipnet1, _ := net.ParseCIDR(test.cidr1)
+		_, ipnet2, _ := net.ParseCIDR(test.cidr2)
+		overlap, err := isOverlap(ipnet1, ipnet2)
+		if err != nil && !test.errorExpected {
+			t.Errorf("test %q failed: got error %s, expected cidr overlap between %s and %s to be %t", test.name, err.Error(), test.cidr1, test.cidr2, test.expected)
+		} else if err == nil && test.errorExpected {
+			t.Errorf("test %q failed: got cidr overlap value %t, expected error", test.name, overlap)
+		} else if !test.errorExpected && overlap != test.expected {
+			t.Errorf("test %q failed: got overlap for cidr %s and %s as %t, expected %t", test.name, test.cidr1, test.cidr2, test.expected, test.expected)
+		}
+	}
+}
+
+func TestIncrementIP(t *testing.T) {
+
+	cases := []struct {
+		name          string
+		currentIP     string
+		step          byte
+		expected      string
+		errorExpected bool
+	}{
+		{
+			name:          "Valid IP increment with step size 143 without carry forward to significant bytes",
+			currentIP:     "192.168.92.32",
+			step:          143,
+			expected:      "192.168.92.175",
+			errorExpected: false,
+		},
+		{
+			name:          "Valid increment with step size 255 and carry forward to significant bytes with maximum step size",
+			currentIP:     "192.255.255.253",
+			step:          255,
+			expected:      "193.0.0.252",
+			errorExpected: false,
+		},
+		{
+			name:          "Valid increment with step size 8 without carry forward to significant bytes",
+			currentIP:     "0.255.255.106",
+			step:          8,
+			expected:      "0.255.255.114",
+			errorExpected: false,
+		},
+		{
+			name:          "Valid increment with step size 8 and carry forward uptil 3rd byte",
+			currentIP:     "0.255.106.255",
+			step:          8,
+			expected:      "0.255.107.7",
+			errorExpected: false,
+		},
+		{
+			name:          "Valid increment with step size 8 and carry forward uptil 2nd byte bytes",
+			currentIP:     "255.106.255.255",
+			step:          8,
+			expected:      "255.107.0.7",
+			errorExpected: false,
+		},
+		{
+			name:          "Valid increment with step size 8 and carry forward uptil 1st byte",
```

Comment:
Can you add in test cases that could potentially catch off-by-one errors?

For example, test cases where the expected should be ".255", ".0" or ".1"

Reply:
i have added the tests